Use std::void_t in HasLegacyDecoder, HasModernDecoder

zdobersek opened the pull request:
#### 3c0f2fdb4126f5bf8adba030e6718fcbc0dc3e23
<pre>
Use std::void_t in HasLegacyDecoder, HasModernDecoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=250275">https://bugs.webkit.org/show_bug.cgi?id=250275</a>

Reviewed by Fujii Hironori.

The HasLegacyDecoder and HasModernDecoder templates used to detect presence of
either decoding method on a given type&apos;s decoding-implementation class are
simplified.

The base templates inherit from std::false_type and gain the third template
parameter that defaults to void. In corresponding partial specializations that
inherit from std::true_type, that third template argument is then specified
as std::void_t instantiation with the type of the desired decoding method.
When the desired method is present, the std::true_type-based partial
specialization will be selected. When not present, the std::false_type-based
base template will be used.

Templated HasLegacyDecoderV and HasModernDecoderV constexpr booleans are
removed as they&apos;re only used in one place.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::ArgumentCoder::decode):
(IPC::Detail::TestLegacyDecoder): Deleted.
(IPC::Detail::TestModernDecoder): Deleted.
* Source/WebKit/Platform/IPC/Decoder.h:

Canonical link: <a href="https://commits.webkit.org/258778@main">https://commits.webkit.org/258778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e798de90ebc0874a151e1b09fe963ef2a1b6660

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11894 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35801 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112031 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172256 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2803 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95026 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109710 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9910 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93120 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37561 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91761 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; 1 api test failed or timed out; re-run-api-tests (cancelled)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24640 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79300 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92989 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5349 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26057 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89341 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/3034 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2503 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29664 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45549 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92263 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6036 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7241 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20665 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WTF/wtf/ArgumentCoder.h:
```diff
@@ -34,16 +34,10 @@ namespace IPC {
+template<typename T, typename I = T, typename = void> struct HasLegacyDecoder : std::false_type { };
+template<typename T, typename I> struct HasLegacyDecoder<T, I, std::void_t<decltype(I::decode(std::declval<Decoder&>(), std::declval<T&>()))>> : std::true_type { };
+template<typename T, typename I = T, typename = void> struct HasModernDecoder : std::false_type { };
+template<typename T, typename I> struct HasModernDecoder<T, I, std::void_t<decltype(I::decode(std::declval<Decoder&>()))>> : std::true_type { };
```
the template argument T is unused. You can remove it.

fujii commented on Source/WTF/wtf/ArgumentCoder.h:
```diff
@@ -34,16 +34,10 @@ namespace IPC {
-template<typename T, typename I = T> struct HasModernDecoder : decltype(Detail::TestModernDecoder<T, I>(0)) { };
-template<typename T, typename I = T> inline constexpr bool HasModernDecoderV = HasModernDecoder<T, I>::value;
+template<typename T, typename I = T, typename = void> struct HasLegacyDecoder : std::false_type { };
+template<typename T, typename I> struct HasLegacyDecoder<T, I, std::void_t<decltype(I::decode(std::declval<Decoder&>(), std::declval<T&>()))>> : std::true_type { };
```
I know why HasLegacyDecoder has two template arguments. But, the second argument isn't used. You can remove it.

zdobersek replied:
Both `T` and `I` are needed due to how `HasModernDecoder` and `HasLegacyDecoder` are used in WebKit, in IPC::Decoder. There, the `I` (as in Impl) is specified as the `ArgumentCoder` specialization, which might be custom-written.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/Platform/IPC/Decoder.h#L95

fujii replied:
Oh sorry, I didn't understand that.

zdobersek committed:
Use std::void_t in HasLegacyDecoder, HasModernDecoder
https://bugs.webkit.org/show_bug.cgi?id=250275

Reviewed by Fujii Hironori.

The HasLegacyDecoder and HasModernDecoder templates used to detect presence of
either decoding method on a given type's decoding-implementation class are
simplified.

The base templates inherit from std::false_type and gain the third template
parameter that defaults to void. In corresponding partial specializations that
inherit from std::true_type, that third template argument is then specified
as std::void_t instantiation with the type of the desired decoding method.
When the desired method is present, the std::true_type-based partial
specialization will be selected. When not present, the std::false_type-based
base template will be used.

Templated HasLegacyDecoderV and HasModernDecoderV constexpr booleans are
removed as they're only used in one place.

* Source/WTF/wtf/ArgumentCoder.h:
(IPC::ArgumentCoder::decode):
(IPC::Detail::TestLegacyDecoder): Deleted.
(IPC::Detail::TestModernDecoder): Deleted.
* Source/WebKit/Platform/IPC/Decoder.h:

Canonical link: https://commits.webkit.org/258778@main
